TL-3.1-HSBM.py

```diff
@@ -0,0 +1,291 @@
+            embed /= row_norms[:, None]
+       
+        return embed
+
+    def _cluster(self, embed, cluster_method=None, n_init=None):
+        if self.cluster_method == "GMM":
+            cluster = AutoGMMCluster(
+                min_components=1, max_components=self.max_clusters
+            )
+            cluster.fit(embed)
+            model = cluster.model_
+            bic_ = model.bic(embed)
+            lik_ = model.score(embed)
+            k = cluster.n_components_
+            pred = cluster.fit_predict(embed)
+
+        res = {
+            "bic": -bic_,
+            "lik": lik_,
+            "k": k,
+            "model": model,
+            "pred": pred,
+        }
+        # TODO: if cluster_method == "Spherical-KMeans":
+        # TODO: if cluster_method == "KMeans":
+
+        return res
+
+    def _model_metrics(self):
+        results = self.results_
+        model = results["model"]
+        pred = results["pred"]
+        _n_components = results["k"]
+
+        return model, pred, _n_components
+
+    def select_model(self, k=None):
+        if k is None:
+            model, pred, k = self._model_metrics()
+
+        self.k_ = k
+        self.children = []
+        if k > 1:
+            self.model_ = model
+            self.pred_ = pred
+            root_labels = self.root.labels
+
+            pred_name = f"{self.depth + 1}_pred"
+            if pred_name not in root_labels.columns:
+                root_labels[pred_name] = ""
+            root_labels.loc[self.root_inds, pred_name] = pred.astype(str)
+
+            uni_labels = np.unique(pred).astype(str)
+
+            self.children = []
+            for i, ul in enumerate(uni_labels):
+                new_labels = root_labels[
+                    (root_labels[pred_name] == ul)
+                    & (root_labels.index.isin(self.root_inds.index))
+                ]
+                new_root_inds = new_labels["inds"]
+                new_name = self.name + "-" + str(ul)
+                new_x = self.root.X[new_root_inds]
+                # new_adj = self.root.adj[np.ix_(new_root_inds, new_root_inds)]
+                if len(new_labels) > self.min_split:
+                    RecursiveCluster(
+                        new_name,
+                        X=new_x,
+                        selection_criteria="bic",
+                        embed_method="ASE",
+                        cluster_method="GMM",
+                        root_inds=new_root_inds,
+                        n_init=1,
+                        reembed=False,
+                        parent=self,
+                        min_clusters=1,
+                        max_clusters=3,
+                        n_components=None,
+                        n_elbows=2,
+                        normalize=False,
+                        regularizer=None,
+                        plus_c=True,
+                        min_split=1,
+                    )
+
+    def collect_labels(self):
```

Comment:
this is a hacky thing that I did. ideally would all be recursive, i think, do you agree? also it may not matter if we get the predict function working like it should